packages/flutter/lib/src/rendering/editable.dart

```diff
@@ -1529,8 +1529,8 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin,
       final Offset start = Offset(0.0, preferredLineHeight) + caretOffset + paintOffset;
       return <TextSelectionPoint>[TextSelectionPoint(start, null)];
     } else {
-      final Offset start = Offset(boxes.first.start, boxes.first.bottom) + paintOffset;
-      final Offset end = Offset(boxes.last.end, boxes.last.bottom) + paintOffset;
+      final Offset start = Offset(boxes.first.start.clamp(0, _textPainter.size.width), boxes.first.bottom) + paintOffset;
+      final Offset end = Offset(boxes.last.end.clamp(0, _textPainter.size.width), boxes.last.bottom) + paintOffset;
```

Comment:
Is this change needed for fixing part 2 or 3? It makes sense to me but I think there're other public APIs that are dependent on `TextPainter.getBoxesForSelection`, so if we change `getEndpointsForSelection` we might want to update those us well to keep the behavior consistent.

Reply:
This change doesn't fix part 2 or 3 directly, but it's required for correct `TextSelectionToolbar` location.
I checked text selection painting and updated it. Now we should update golden tests images. I described in the https://github.com/flutter/flutter/pull/91698#discussion_r738764737 that `letterSpacing` can bring the letters closer, but it can also move caret/selection outside (by value from `getRectsForRange`).
I will also check other items that are depends on `getBoxesForSelection`.